server/src/main/java/org/elasticsearch/index/seqno/RetentionLease.java

```diff
@@ -81,18 +88,118 @@ public String source() {
      * @param source                  the source of the retention lease
      */
     public RetentionLease(final String id, final long retainingSequenceNumber, final long timestamp, final String source) {
+        Objects.requireNonNull(id);
+        if (id.isEmpty()) {
+            throw new IllegalArgumentException("retention lease ID can not be empty");
+        }
+        if (id.contains(":") || id.contains(";") || id.contains(",")) {
+            // retention lease IDs can not contain these characters because they are used in encoding retention leases
+            throw new IllegalArgumentException("retention lease ID can not contain any of [:;,] but was [" + id + "]");
+        }
         if (retainingSequenceNumber < SequenceNumbers.UNASSIGNED_SEQ_NO) {
             throw new IllegalArgumentException("retention lease retaining sequence number [" + retainingSequenceNumber + "] out of range");
         }
         if (timestamp < 0) {
             throw new IllegalArgumentException("retention lease timestamp [" + timestamp + "] out of range");
         }
+        Objects.requireNonNull(source);
+        if (source.isEmpty()) {
+            throw new IllegalArgumentException("retention lease source can not be empty");
+        }
+        if (source.contains(":") || source.contains(";") || source.contains(",")) {
+            // retention lease sources can not contain these characters because they are used in encoding retention leases
+            throw new IllegalArgumentException("retention lease source can not contain any of [:;,] but was [" + source + "]");
+        }
         this.id = id;
         this.retainingSequenceNumber = retainingSequenceNumber;
         this.timestamp = timestamp;
         this.source = source;
     }
 
+    /**
+     * Encodes a retention lease as a string. This encoding can be decoded by {@link #decodeRetentionLease(String)}. The retention lease is
+     * encoded in the format <code>id:{id};retaining_seq_no:{retainingSequenecNumber};timestamp:{timestamp};source:{source}</code>.
+     *
+     * @param retentionLease the retention lease
+     * @return the encoding of the retention lease
+     */
+    static String encodeRetentionLease(final RetentionLease retentionLease) {
+        Objects.requireNonNull(retentionLease);
+        return String.format(
+                Locale.ROOT,
+                "id:%s;retaining_seq_no:%d;timestamp:%d;source:%s",
```

Comment:
I wonder if we should use JSON, with the upside of being more extensible and will allow us to use our parsing infra (and the validation features it has). WDYT?

Reply:
I am not particularly concerned about this needing to be extensible. I can be convinced that JSON is worth it but I am doubting.